base/broadcast.jl

```diff
@@ -991,11 +991,17 @@ preprocess_args(dest, args::Tuple{}) = ()
 # Specialize this method if all you want to do is specialize on typeof(dest)
 @inline function copyto!(dest::AbstractArray, bc::Broadcasted{Nothing})
     axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
-    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match
+    # Performance optimization: broadcast!(identity, dest, A) is equivalent to copyto!(dest, A) if indices match.
+    # However copyto!(dest, A) is very slow in many cases, implement a faster version here.
```

Comment:
Wouldn't it be better to speed up `copyto!(dest, A)`?

Reply:
In fact, I tried, but the problem seems to be a little complicated with the following facts:

1. `copyto!(::Array,::Array)` call C's `memmove`, but it seems slower than a single loop:
```julia
julia> @btime copyto!($b,$a);                    # this call C's memmove
  120.524 ns (0 allocations: 0 bytes)
julia> @btime copyto!(IndexLinear(),$b,IndexLinear(),$a);  # this call Julia's copyto_unalias!
  74.512 ns (0 allocations: 0 bytes)
```
2.  For `abstractarray`, I can't find a way to accelerate `copyto_unalias!(dest, A)` without `@simd`, as:
```julia
a = randn(40,40); b = similar(a);
f1!(a,b) = @inbounds for i in eachindex(IndexCartesian(), a)
    b[i] = a[i]
end

f2!(a,b) = @inbounds @simd for i in eachindex(IndexCartesian(), a)
    b[i] = a[i]
end
@btime f1!($a, $b)   #247.132 ns (0 allocations: 0 bytes)
@btime f2!($a, $b)  #220.548 ns (0 allocations: 0 bytes)  a little faster
@btime @views f1!($a[1:end,1:end], $b[1:end,1:end]) #   1.710 μs (0 allocations: 0 bytes)
@btime @views f2!($a[1:end,1:end], $b[1:end,1:end]) #   262.500 ns (0 allocations: 0 bytes) much faster
```

Reply:
Why can't you use `@simd`?

Reply:
The buildbot throw an error that ’@simd‘ is not defined. I‘m not familiar with the build system, but it seems to be a world age problem?(abstractarray.jl is included before simd.jl in Bass.jl)

Reply:
Sounds just like a bootstrapping problem, i.e. `@simd` is not defined until later in the build process.  That should be fixable by e.g. by moving the `copyto!` definition to a file loaded later.